File: src/minimcp/time_limiter.py
import logging
from typing import Any

from anyio import CancelScope, current_time

logger = logging.getLogger(__name__)


class TimeLimiter:
    """Enforces an idle timeout for a single message handler.

    The timer can be reset during handler execution to extend the deadline,
    preventing timeout when the handler is actively processing.

    Used as a synchronous context manager — the cancel scope is entered on
    ``__enter__`` and exited on ``__exit__``.
    """

    _timeout: float
    _scope: CancelScope

    def __init__(self, timeout: float):
        """
        Args:
            timeout: The idle timeout in seconds.
        """
        self._timeout = float(timeout)
        self._scope = CancelScope()
        self.reset()

    def reset(self) -> None:
        """Reset the idle timeout, extending the deadline from the current time."""
        self._scope.deadline = current_time() + self._timeout

    def __enter__(self):
        self._scope.__enter__()
        return self

    def __exit__(self, *args: Any):
        return self._scope.__exit__(*args)

File: rust/crates/minimcp/src/time_limiter.rs
//! Idle timeout management for a single message handler.
//!
//! Mirrors `minimcp/time_limiter.py`. In the Python implementation the limiter
//! wraps the handler in an anyio cancel scope whose deadline can be reset. In
//! this skeleton [`crate::MiniMCP::handle`] applies the timeout via
//! `tokio::time::timeout`; `TimeLimiter` is the public handle exposed to
//! handlers for resetting the deadline.

use std::time::Duration;

/// Enforces an idle timeout for a single message handler.
///
/// The deadline can be extended via [`TimeLimiter::reset`] while the handler is
/// actively processing (for example, while streaming progress notifications).
#[derive(Debug, Clone)]
pub struct TimeLimiter {
    timeout: Duration,
}

impl TimeLimiter {
    /// Create a new limiter with the given idle timeout.
    pub fn new(timeout: Duration) -> Self {
        Self { timeout }
    }

    /// The configured idle timeout.
    pub fn timeout(&self) -> Duration {
        self.timeout
    }

    /// Reset the idle deadline, extending it from the current time.
    ///
    /// TODO: Wire this to the active timeout future so resets actually extend
    /// the deadline. The skeleton currently applies a fixed per-request timeout.
    pub fn reset(&self) {
        tracing::trace!("TimeLimiter::reset called (no-op in skeleton)");
    }
}
